### ISSUE TITLE
Problems with PPO on MiniGrid

### ISSUE THREAD

**Issue by ErikVester** [DLR-RM/stable-baselines3#324]
Hi,

This is something I have been working on the past 2 weeks and I am starting to become convinced that something is actually going wrong here, however I still did not manage to exactly pinpoint what causes this behaviour! 

### Describe the bug

Main: 
Training suddenly collapses in PPO when training on MiniGrid environment. This continuously seems to happen as the policy approaches/becomes the optimal policy. Normalising seems to help, but doesn't take it away. I have tried this for all kinds of different hyperparameters and it keeps happening. Lowering the learning rate makes it more stable but doesn't solve the problem of exploding KL divergence and explained variance becoming negative. See plots below: 
![image](https://user-images.githubusercontent.com/45073001/108627871-f9fcc080-7457-11eb-8f8c-531f4352400f.png)

As far as I know this behaviour should not happen. Furthermore, this behaviour does not happen when I use for instance A2C. I wasn't able to find any mistakes in the environment, which also passes the env_checker. 
I find it very difficult to pinpoint what the exact thing is that goes wrong, any ideas what I could still check/do to figure out where the problem actually lies? Any help would be appreciated! I am definitely willing to put in the effort to figure out if this is a problem in my env, just PPO behaviour or indeed a bug, but I am just not sure how to further approach this. (I did try to log/print things, but so far that didn't give me any helpful insights, apart from the fact the agent seems to locks onto one specific action when) 

### Code example
As I am using MiniGrid here, you will first have to install it: pip3 install gym-minigrid

```python
import gym_minigrid
import gym
from gym import spaces

from stable_baselines3 import PPO

from stable_baselines3.ppo import MlpPolicy

from stable_baselines3.common.env_checker import check_env
from stable_baselines3.common.evaluation import evaluate_policy
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.utils import set_random_seed
from stable_baselines3.common.callbacks import EvalCallback

from gym_minigrid.wrappers import *



# The environment to train on
env_name = 'MiniGrid-Empty-6x6-v0'

# The name of the run  ## TODO: make this variable 
test_name = 'bug_test'

#  Parallel environments
env = make_vec_env(env_name, wrapper_class= FlatObsWrapper, n_envs= 10)

# Separate env for evaluation
eval_env = make_vec_env(env_name, wrapper_class= FlatObsWrapper, n_envs=1)

# Checking env
env_check = gym.make(env_name)
env_check = FlatObsWrapper(env_check)
check_env(env_check)

# Evaluation callback
eval_callback = EvalCallback(eval_env, best_model_save_path='./saved_models/'+test_name+'/',
                             log_path='./logs/', eval_freq=1000,
                             deterministic= False, render=False, verbose = 0)

policy_kwargs = dict(normalize_images = True, net_arch = [128, dict(vf=[256], pi=[16])])

# Reproducable
seed = 0
env.seed(seed)
env.action_space.seed(seed)
set_random_seed(seed)

# Learning the model
model = PPO(MlpPolicy, env ,policy_kwargs= policy_kwargs, verbose=1, learning_rate= 0.0003, n_steps = 208, batch_size= 65,  tensorboard_log='./saved_models/'+test_name+'/tensorboard/', seed = seed) 
model.learn(total_timesteps=120000, callback= eval_callback)
model.save('saved_models/'+ test_name+'/'+'final')
```

### System Info
Describe the characteristic of your environment:

 * Library installed from source
 * Running on CPU
 * Python version: 3.8.7
 * PyTorch version: 1.7.1
 * Gym version: 0.18.0
 * MiniGrid installation: pip3 install gym-minigrid

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [x] I have checked my env using the env checker (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by araffin** [comment]
>Training suddenly collapses in PPO when training on MiniGrid environment.
>This continuously seems to happen as the policy approaches/becomes the optimal policy. 
>See plots below:

Looking at your plots, it seems that PPO learns the optimal policy, collapses a bit and then converges back to the optimal one, no?
(also I do not see the evalcallback graph, which is usually the most important one).

This happens on some environments (for instance HopperBullet), and in those cases, you can usually do two things:
- use a linearly decreasing clip range: see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L227 to make the update more conservative over time
- use a linearly decreasing learning rate, see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L11

Anyway, I still recommend you to use the rl zoo and possibly automatic hyperparameter tuning (included in the zoo).

**Comment by ErikVester** [comment]
> > Training suddenly collapses in PPO when training on MiniGrid environment.
> > This continuously seems to happen as the policy approaches/becomes the optimal policy.
> > See plots below:
> 
> Looking at your plots, it seems that PPO learns the optimal policy, collapses a bit and then converges back to the optimal one, no?
> (also I do not see the evalcallback graph, which is usually the most important one).
> 
> This happens on some environments (for instance HopperBullet), and in those cases, you can usually do two things:
> 
> * use a linearly decreasing clip range: see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L227 to make the update more conservative over time
> * use a linearly decreasing learning rate, see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L11
> 
> Anyway, I still recommend you to use the rl zoo and possibly automatic hyperparameter tuning (included in the zoo).

Hi Araffin, 
thank you for your reply again, said it before, but really appreciate the effort you guys put in this project and also in helping people out! I waited a bit with a reply as I first wanted to test some things!
Still some findings: 
> Looking at your plots, it seems that PPO learns the optimal policy, collapses a bit and then converges back to the optimal one, no? 
Sometimes is also completely collapses. 

One of the things that really seemed to help is just simply adding some more randomness to the environment through random initialisation. It seems as if the optimal policy results in such a short episode, that the agent gets to see very little different states at one point which might cause this unstable behaviour. I understood now that this is something more common for neural nets with too little variance in the input.

**Comment by ErikVester** [comment]
Hi Araffin, 
thank you for your reply again, I said it before, but I really appreciate the effort you guys put in this project and also in helping people out! I waited a bit with a reply as I first wanted to test some things!

Still some findings: 
One of the things that really seemed to help is just simply adding some more randomness to the environment through random initialisation. It seems as if the optimal policy results in such a short episode, that the agent gets to see very little different states at one point which might cause this unstable behaviour. I understood now that this is something more common for neural nets with too little variance in the input. 

> Looking at your plots, it seems that PPO learns the optimal policy, collapses a bit and then converges back to the optimal one, no? 

Sometimes it also completely collapsed to zero and stayed there. 

> (also I do not see the evalcallback graph, which is usually the most important one).

They look very similar, except a lot more 'spiky' when I use deterministic actions. In general this remains, even when the algorithm does seem stable during training now: 
![image](https://user-images.githubusercontent.com/45073001/109954103-30043500-7ce1-11eb-9e5f-0bcb1419c937.png)

Is this normal?


> * use a linearly decreasing clip range: see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L227 to make the update more conservative over time
> * use a linearly decreasing learning rate, see https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L11

Thanks! I actually already played around with this and it does help, I just wasn't sure if something was actually wrong and if this would  then be considered as 'brushing over' instead of solving the underlying problem. 

>Anyway, I still recommend you to use the rl zoo and possibly automatic hyperparameter tuning (included in the zoo).

I also tried this and it works very well for tuning, but not for making the algorithm more stable  - unstable being the behaviour I described above - as the objective function on which the algorithm is tuned is solely dependent on reward, so a setting which is unstable but did end with good reward is considered good. 

Kind regards, 
Erik

**Comment by BaiYunpeng1949** [comment]
Hi, 

I have some similar problems. I am using a MuJoCo+gym custom environment with stb3-PPO running 16 envs in parallel. 

[My tensorboard logs](https://tensorboard.dev/experiment/7XgKjENUQpiVXrIwzlWXkg/#scalars).

My ep_rew_mean always drops sharply close to 0 after a period of training time, I call this timestep as **the collapse point**. I noticed that at the collapse point, all other training metrics, such as approx_kl, loss, entropy_loss, policy_gradient_loss are having a outstanding spike / abnormal value. Is this something similar to catastrophic forgetting? Or it is just due to some bugs in my custom env?

![Screenshot 2023-04-06 123540](https://user-images.githubusercontent.com/54785878/230273247-88d6a443-1bad-486d-a18b-99a07ea8b77c.jpg)

![image](https://user-images.githubusercontent.com/54785878/230273349-5b067467-0f46-4cbe-a6eb-383985fbdcb4.png)


Most of my PPO hyperparameters are default values. Just the 
`num_steps=5000`
`batch_size=500`

Can you help me have a quick check of that? @araffin  Many thanks! 

Best regards
Bai
